server/src/main/java/org/opensearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -167,6 +174,42 @@ public static long sizeOfRelocatingShards(
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
         ClusterInfo clusterInfo = allocation.clusterInfo();
+
+        /*
+         The following block enables allocation for remote shards within safeguard limits of the filecache.
+         */
+        if (REMOTE_CAPABLE.equals(getNodePool(node)) && REMOTE_CAPABLE.equals(getShardPool(shardRouting, allocation))) {
+            final Predicate<ShardRouting> shardRoutingPredicate = shard -> shard.primary()
+                && REMOTE_CAPABLE.equals(getShardPool(shard, allocation));
+            final List<ShardRouting> remoteShardsOnNode = shardsSatisfyingPredicate(node, shardRoutingPredicate);
+            final long currentNodeRemoteShardSize = remoteShardsOnNode.stream()
+                .map(ShardRouting::getExpectedShardSize)
+                .mapToLong(Long::longValue)
+                .sum();
+
+            final long shardSize = getExpectedShardSize(
+                shardRouting,
+                0L,
+                allocation.clusterInfo(),
+                allocation.snapshotShardSizeInfo(),
+                allocation.metadata(),
+                allocation.routingTable()
+            );
+
+            final FileCacheStats fileCacheStats = clusterInfo.getNodeFileCacheStats().getOrDefault(node.nodeId(), null);
+            final long nodeCacheSize = fileCacheStats != null ? fileCacheStats.getTotal().getBytes() : 0;
+            final long totalNodeRemoteShardSize = currentNodeRemoteShardSize + shardSize;
+
+            if (totalNodeRemoteShardSize > DATA_TO_FILE_CACHE_SIZE_RATIO * nodeCacheSize) {
```

Comment:
For backward compatibility I think this has to be an opt-in setting (or the default ratio set to be implausibly huge so as to never be a limit), otherwise its a breaking change.

Reply:
Yep. I have that on my plan listed out on the original PR.
I will follow up with a BWC config as well as changes to the `RestoreService` as on the previous PR. 

https://github.com/opensearch-project/OpenSearch/pull/7713#issuecomment-1601824927